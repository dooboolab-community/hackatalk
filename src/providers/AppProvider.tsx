import React, { useReducer } from 'react';

import { AppContext } from '../contexts';
import { ThemeType } from '../theme';
import { User } from '../types';

const AppConsumer = AppContext.Consumer;

interface Action {
  type: 'reset-user' | 'set-user' | 'change-theme-mode';
  payload: any;
}

interface Props {
  navigation?: any;
  children?: any;
}

export interface State {
  user: User;
  theme: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
  user: {
    displayName: '',
    age: 0,
    job: '',
  },
};

const reducer = (state: State, action: Action) => {
  switch (action.type) {
  case 'change-theme-mode':
    return { ...state, theme: action.payload.theme };
  case 'reset-user':
    return { ...state, user: initialState.user };
  case 'set-user':
    return { ...state, user: action.payload };
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = { state, dispatch };

  return (
    <AppContext.Provider value={value}>{props.children}</AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
