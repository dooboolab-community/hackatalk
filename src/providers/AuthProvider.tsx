import React, { useReducer } from 'react';

import Relay from '../relay';
import { Environment as RelayEnvironment } from 'react-relay';
import { User } from '../types/graphql';
import createCtx from '../utils/createCtx';

interface Context {
  state: State;
  setUser(user: User | undefined): void;
}
const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  SetUser = 'set-user',
}

export interface State {
  user: User | undefined | null;
  relay: RelayEnvironment;
}
type SetUserAction = {
  type: ActionType.SetUser;
  payload: { user: User; relay: RelayEnvironment };
};
type Action = SetUserAction;

interface Props {
  children?: React.ReactElement;
  initialAuthUser?: User;
}

type Reducer = (state: State, action: Action) => State;

const setUser = (dispatch: React.Dispatch<SetUserAction>) => (
  authUser: User,
): void => {
  if (!authUser) Relay.init();
  dispatch({
    type: ActionType.SetUser,
    payload: { user: authUser, relay: Relay.environment },
  });
};

const initialState: State = {
  user: undefined,
  relay: Relay.environment,
};

const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SetUser:
      return { ...action.payload };
    default:
      return state;
  }
};

function AuthProvider(props: Props): React.ReactElement {
  const initialAuthState = {
    user: props.initialAuthUser,
    relay: Relay.environment,
  };
  const [state, dispatch] = useReducer<Reducer>(reducer, initialAuthState);

  const actions = {
    setUser: setUser(dispatch),
  };

  return <Provider value={{ state, ...actions }}>{props.children}</Provider>;
}

const AuthContext = {
  useAuthContext: useCtx,
  AuthProvider,
};

export { useCtx as useAuthContext, AuthProvider };
export default AuthContext;
