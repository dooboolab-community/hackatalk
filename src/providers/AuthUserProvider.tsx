import React, { useReducer } from 'react';

import { AuthUser } from '../types';
import createCtx from '../utils/createCtx';

interface Context {
  state: State;
  setAuthUser(authUser: AuthUser): void;
}
const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  SetAuthUser = 'set-auth-user',
}

export interface State {
  user?: AuthUser;
}

type Action =
  | { type: ActionType.SetAuthUser; payload: State };

interface Props {
  children?: React.ReactElement;
  initialAuthUser?: AuthUser;
}

type Reducer = (state: State, action: Action) => State;

const setAuthUser = (dispatch: React.Dispatch<Action>) => (authUser: AuthUser): void => {
  dispatch({
    type: ActionType.SetAuthUser,
    payload: { user: authUser },
  });
};

const initialState: State = {};

const reducer: Reducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case ActionType.SetAuthUser:
      return {
        user: payload.user,
      };
    default:
      return state;
  }
};

function AuthUserProvider(props: Props): React.ReactElement {
  const { children, initialAuthUser } = props;
  const [state, dispatch] = useReducer<Reducer>(reducer, { user: initialAuthUser });

  const actions = {
    setAuthUser: setAuthUser(dispatch),
  };

  return <Provider value={{ state, ...actions }}>{children}</Provider>;
}

export { useCtx as useAuthUserContext, AuthUserProvider };
