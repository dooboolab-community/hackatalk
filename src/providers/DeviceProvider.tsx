import * as Device from 'expo-device';

import React, { useState } from 'react';

import createCtx from '../utils/createCtx';

interface Context {
  deviceType: Device.DeviceType;
  setDeviceType: React.Dispatch<React.SetStateAction<Device.DeviceType>>;
}
const [useCtx, Provider] = createCtx<Context>();

interface Props {
  initialDeviceType?: Device.DeviceType;
  children?: React.ReactElement;
}

function DeviceProvider(props: Props): React.ReactElement {
  const { initialDeviceType = Device.DeviceType.PHONE, children } = props;
  const [deviceType, setDeviceType] = useState<Device.DeviceType >(initialDeviceType);

  return (
    <Provider
      value={{
        deviceType,
        setDeviceType,
      }}
    >
      {children}
    </Provider>
  );
}

const DeviceContext = {
  useDeviceContext: useCtx,
  DeviceProvider,
};

export { useCtx as useDeviceContext, DeviceProvider };
export default DeviceContext;
