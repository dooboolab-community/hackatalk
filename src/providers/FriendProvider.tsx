import React, { useReducer } from 'react';

import { User } from '../types';
import createCtx from '../utils/createCtx';
import produce from 'immer';

interface Context {
  friendState: State;
  addFriend: (user: User) => void;
  deleteFriend: (user: User) => void;
  setFriends: (users: User[]) => void;
}

const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  AddFriend = 'add-friend',
  DeleteFriend = 'delete-friend',
  SetFriends = 'set-friends'
}

export interface State {
  friends: User[];
}

interface Payload {
  user?: User;
  users?: User[];
}

const MockInitialFriends = [
  {
    id: '0',
    nickname: 'hello',
    thumbURL: '',
    photoURL: '',
    statusMessage: 'I am fine today',
    isOnline: true,
  },
];

const initialState: State = {
  friends: [...MockInitialFriends],
};

type Action = { type: ActionType; payload: Payload };

interface Props {
  children?: React.ReactElement;
}

type Reducer = (state: State, action: Action) => State;

const addFriend = (dispatch: React.Dispatch<Action>) => (user: User): void => {
  dispatch({
    type: ActionType.AddFriend,
    payload: { user },
  });
};

const deleteFriend = (dispatch: React.Dispatch<Action>) => (
  user: User,
): void => {
  dispatch({
    type: ActionType.DeleteFriend,
    payload: { user },
  });
};

const setFriends = (dispatch: React.Dispatch<Action>) => (
  users: User[],
): void => {
  dispatch({
    type: ActionType.SetFriends,
    payload: { users },
  });
};

const reducer: Reducer = (state = initialState, action) => {
  return produce(state, (draft) => {
    const { type, payload } = action;
    switch (type) {
      case ActionType.AddFriend: {
        const { user } = payload;
        if (user && !draft.friends.find((friend) => friend.id === user.id)) {
          const index = draft.friends.findIndex(
            (friend) =>
              (user.nickname || '').toLowerCase() <
              (friend?.nickname || '').toLowerCase(),
          );
          draft.friends.splice(
            index === -1 ? draft.friends.length : index,
            0,
            user,
          );
        }
        break;
      }
      case ActionType.DeleteFriend: {
        const { user } = payload;
        if (user) {
          const index = draft.friends.findIndex(
            (friend) => friend.id === user.id,
          );
          if (index !== -1) {
            draft.friends.splice(index, 1);
          }
        }
        break;
      }
      case ActionType.SetFriends: {
        const { users } = payload;
        if (users) {
          draft.friends = users;
        }
        break;
      }
    }
  });
};

function FriendProvider(props: Props): React.ReactElement {
  const [friendState, dispatch] = useReducer<Reducer>(reducer, initialState);

  const actions = {
    // legacy
    addFriend: addFriend(dispatch),
    deleteFriend: deleteFriend(dispatch),
    setFriends: setFriends(dispatch),
  };
  return (
    <Provider value={{ friendState, ...actions }}>{props.children}</Provider>
  );
}

export { useCtx as useFriendContext, FriendProvider };
