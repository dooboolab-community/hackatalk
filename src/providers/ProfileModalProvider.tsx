import React, { MutableRefObject, useReducer } from 'react';

import { Ref as ProfileModalRef } from '../components/shared/ProfileModal';
import { User } from '../types';
import createCtx from '../utils/createCtx';

interface ShowModalParams {
  user: User;
  deleteMode: boolean;
  onDeleteFriend?: () => void;
  onAddFriend?: () => void;
}

interface Context {
  state: State;
  showModal: (showModalParams: ShowModalParams) => void;
  // setUser: (user: User) => void;
  // setShowAddBtn: (deleteMode: boolean) => void;
  // setScreen: (screen: string) => void;
  // open: () => void;
}

const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  ShowModal = 'show-modal',
}

export interface State {
  user: User;
  deleteMode: boolean;
  modal?: React.MutableRefObject<ProfileModalRef | null>;
}

export interface Payload extends State {
  onDeleteFriend?: () => void;
  onAddFriend?: () => void;
}

const initialState: State = {
  user: {
    id: '',
    nickname: '',
    photoURL: '',
    statusMessage: '',
  },
  deleteMode: false,
  modal: undefined,
};

type Action = { type: ActionType.ShowModal; payload: Payload };

interface Props {
  children?: React.ReactElement;
}

type Reducer = (state: State, action: Action) => State;

const showModal = (dispatch: React.Dispatch<Action>) => ({
  user,
  deleteMode,
  onDeleteFriend,
  onAddFriend,
}: ShowModalParams): void => {
  dispatch({
    type: ActionType.ShowModal,
    payload: {
      user,
      deleteMode,
      onDeleteFriend,
      onAddFriend,
    },
  });
};

const reducer: Reducer = (state = initialState, action) => {
  const { type, payload } = action;
  const { modal } = state;
  switch (type) {
    case ActionType.ShowModal:
      if (modal && modal.current) {
        modal.current.setUser(payload.user);
        modal.current.showAddBtn(!payload.deleteMode);
        modal.current.open();
      }
      return {
        ...state,
        user: payload.user,
        deleteMode: !payload.deleteMode,
      };
    default:
      return state;
  }
};

function ProfileModalProvider(props: Props): React.ReactElement {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);

  const actions = {
    showModal: showModal(dispatch),
  };

  return <Provider value={{ state, ...actions }}>{props.children}</Provider>;
}

const ProfileContext = {
  useProfileContext: useCtx,
  ProfileModalProvider,
};

export { useCtx as useProfileContext, ProfileModalProvider };
export default ProfileContext;
