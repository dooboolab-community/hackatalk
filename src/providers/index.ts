import { AppConsumer, AppContext, AppProvider } from './AppProvider';

export {
  AppProvider,
  AppConsumer,
  AppContext,
};
