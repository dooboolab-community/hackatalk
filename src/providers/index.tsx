import * as Device from 'expo-device';

import { ThemeProvider, ThemeType } from '@dooboo-ui/native-theme';
import { dark, light } from '../theme';

import { ApolloProvider } from '@apollo/react-hooks';
import { AuthProvider } from './AuthProvider';
import { DeviceProvider } from './DeviceProvider';
import { FriendProvider } from './FriendProvider';
import { ProfileModalProvider } from './ProfileModalProvider';
import React from 'react';
import { User } from '../types';
import testClient from '../apollo/testClient';

interface Props {
  initialDeviceType?: Device.DeviceType;
  initialThemeType?: ThemeType;
  initialAuthUser?: User;
  children?: React.ReactElement;
}

// hyochan => for testing
export const AllProviders = ({
  initialThemeType,
  initialAuthUser,
  initialDeviceType,
  children,
}: Props): React.ReactElement => {
  return (
    <DeviceProvider
      initialDeviceType={initialDeviceType}
    >
      <ThemeProvider
        initialThemeType={initialThemeType}
        customTheme={{ light, dark }}
      >
        <ApolloProvider client={testClient}>
          <FriendProvider>
            <AuthProvider initialAuthUser={initialAuthUser}>
              <ProfileModalProvider>
                {children}
              </ProfileModalProvider>
            </AuthProvider>
          </FriendProvider>
        </ApolloProvider>
      </ThemeProvider>
    </DeviceProvider>
  );
};
