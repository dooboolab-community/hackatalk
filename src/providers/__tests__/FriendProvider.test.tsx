import * as React from 'react';

import { Button, FlatList, View } from 'react-native';
import {
  RenderResult,
  act,
  fireEvent,
  render,
} from '@testing-library/react-native';

import { AllProviders } from '../../providers';
import { User } from '../../types';
import UserListItem from '../../components/shared/UserListItem';
import { useFriendContext } from '../FriendProvider';

const fakeUsers: User[] = [
  {
    id: '1',
    nickname: 'admin',
    thumbURL: 'https://avatars2.githubusercontent.com/u/45788556?s=200&v=4',
    photoURL: 'https://avatars2.githubusercontent.com/u/45788556?s=200&v=4',
    statusMessage: 'online',
    isOnline: true,
  },
  {
    id: '2',
    nickname: 'zeoseong',
    thumbURL: 'https://avatars2.githubusercontent.com/u/19166187?s=460&v=4',
    photoURL: 'https://avatars2.githubusercontent.com/u/19166187?s=460&v=4',
    statusMessage: 'hello',
    isOnline: false,
  },
];

const FakeChild = (): React.ReactElement => {
  const {
    friendState: { friends },
    addFriend,
    deleteFriend,
  } = useFriendContext();

  const renderItem = ({ item }): React.ReactElement => (
    <UserListItem user={item} />
  );

  return (
    <View testID="fake-child">
      <Button
        testID="btn-add-1"
        title="add"
        onPress={(): void => addFriend(fakeUsers[0])}
      />
      <Button
        testID="btn-add-2"
        title="add"
        onPress={(): void => addFriend(fakeUsers[1])}
      />
      <Button
        testID="btn-delete-1"
        title="delete"
        onPress={(): void => deleteFriend(fakeUsers[0])}
      />
      <Button
        testID="btn-delete-2"
        title="delete"
        onPress={(): void => deleteFriend(fakeUsers[1])}
      />
      <FlatList
        testID="friend-list"
        keyExtractor={(item): string => `${item.id}`}
        data={friends}
        renderItem={renderItem}
      />
    </View>
  );
};

const TestComponent = (): React.ReactElement => (
  <AllProviders>
    <FakeChild />
  </AllProviders>
);

describe('[FriendProvider] rendering test', () => {
  it('component and snapshot matches', () => {
    const { baseElement } = render(<TestComponent/>);
    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});

describe('[FriendProvider] interactions', () => {
  let testingLib: RenderResult;
  let initFriendList: Partial<User>[];
  let initFriendCount: number;

  beforeEach(() => {
    testingLib = render(<TestComponent />);
    const friendList = testingLib.queryByTestId('friend-list');
    initFriendList = [...friendList.props.data];
    initFriendCount = initFriendList.length;
  });

  describe('[FriendProvider] add friends', () => {
    it('should be added only once, when add the same user twice', () => {
      const btnAdd1 = testingLib.queryByTestId('btn-add-1');

      //
      act(() => {
        fireEvent.press(btnAdd1);
      });
      let friendList = testingLib.queryByTestId('friend-list');
      expect(friendList.props.data.length).toBe(initFriendCount + 1);
      const registeredFriendList = [
        ...initFriendList,
        fakeUsers[0],
      ].sort((a, b) => (a.nickname > b.nickname ? 1 : -1));
      expect(friendList.props.data).toMatchObject(registeredFriendList);

      //
      act(() => {
        fireEvent.press(btnAdd1);
      });
      friendList = testingLib.queryByTestId('friend-list');
      expect(friendList.props.data.length).toBe(initFriendCount + 1);
      expect(friendList.props.data).toMatchObject(registeredFriendList);
    });

    it('should be added 2 users, when added 2 unregistered users', () => {
      const btnAdd1 = testingLib.queryByTestId('btn-add-1');
      const btnAdd2 = testingLib.queryByTestId('btn-add-2');
      act(() => {
        fireEvent.press(btnAdd1);
        fireEvent.press(btnAdd2);
      });
      const friendList = testingLib.queryByTestId('friend-list');
      const registeredFriendList = [
        ...initFriendList,
        fakeUsers[0],
        fakeUsers[1],
      ].sort((a, b) => (a.nickname > b.nickname ? 1 : -1));
      expect(friendList.props.data.length).toBe(initFriendCount + 2);
      expect(friendList.props.data).toMatchObject(registeredFriendList);
    });
  });

  describe('[FriendProvider] delete friends', () => {
    // add friends
    beforeEach(() => {
      const btnAdd1 = testingLib.queryByTestId('btn-add-1');
      const btnAdd2 = testingLib.queryByTestId('btn-add-2');
      act(() => {
        fireEvent.press(btnAdd1);
        fireEvent.press(btnAdd2);
      });
      const friendList = testingLib.queryByTestId('friend-list');
      initFriendList = friendList.props.data;
      initFriendCount = initFriendList.length;
    });

    it('should be deleted, when delete registered friend', () => {
      const btnDelete1 = testingLib.queryByTestId('btn-delete-1');

      //
      act(() => {
        fireEvent.press(btnDelete1);
      });
      let friendList = testingLib.queryByTestId('friend-list');
      expect(friendList.props.data.length).toBe(initFriendCount - 1);

      //
      act(() => {
        fireEvent.press(btnDelete1);
      });
      friendList = testingLib.queryByTestId('friend-list');
      expect(friendList.props.data.length).toBe(initFriendCount - 1);
    });

    it('should be deleted 2 user, when delete 2 registered friends', () => {
      const btnDelete1 = testingLib.queryByTestId('btn-delete-1');
      const btnDelete2 = testingLib.queryByTestId('btn-delete-2');

      //
      act(() => {
        fireEvent.press(btnDelete1);
        fireEvent.press(btnDelete2);
      });
      const friendList = testingLib.queryByTestId('friend-list');
      expect(friendList.props.data.length).toBe(initFriendCount - 2);
    });
  });
});
