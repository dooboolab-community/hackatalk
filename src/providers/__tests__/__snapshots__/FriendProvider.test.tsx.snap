// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[FriendProvider] rendering test component and snapshot matches 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      testID="fake-child"
    >
      <Button
        testID="btn-add-1"
        title="add"
      />
      <Button
        testID="btn-add-2"
        title="add"
      />
      <Button
        testID="btn-delete-1"
        title="delete"
      />
      <Button
        testID="btn-delete-2"
        title="delete"
      />
      <ScrollView
        data={
          Array [
            Object {
              "id": "0",
              "isOnline": true,
              "nickname": "hello",
              "photoURL": "",
              "statusMessage": "I am fine today",
              "thumbURL": "",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        testID="friend-list"
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "width": "100%",
                },
              ]
            }
          >
            <TouchableOpacity
              activeOpacity={0.5}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#ffffff",
                      "borderBottomWidth": 1,
                      "borderColor": "#e2f2ff",
                      "flexDirection": "row",
                      "height": 80,
                      "justifyContent": "flex-start",
                      "paddingBottom": 0,
                      "paddingLeft": 20,
                      "paddingRight": 20,
                      "paddingTop": 0,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "height": 40,
                      "justifyContent": "center",
                      "width": 40,
                    }
                  }
                >
                  <Text />
                </View>
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#151A25",
                        "fontSize": 14,
                        "marginLeft": 12,
                        "width": 200,
                      },
                    ]
                  }
                >
                  hello
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "borderColor": "#d3d8e8",
                        "borderWidth": 0.3,
                        "color": "#1E6EFA",
                        "fontSize": 12,
                        "maxWidth": 120,
                        "paddingBottom": 4,
                        "paddingLeft": 8,
                        "paddingRight": 8,
                        "paddingTop": 4,
                        "position": "absolute",
                        "right": 20,
                      },
                    ]
                  }
                >
                  I am fine today
                </Text>
              </View>
            </TouchableOpacity>
          </View>
        </View>
      </ScrollView>
    </View>
  </View>
</View>
`;
